File: android/TargetingSignals.kt
package com.dkmads.ssp

/**
 * Structured publisher targeting payload for bid `signals` and optional FPD sync.
 * Field names align with server/lib/targeting-signals.js and docs/TARGETING_SIGNALS.md.
 */
data class TargetingSignals(
    val userPid: String? = null,
    val devicePid: String? = null,
    val gender: String? = null,
    val age: Int? = null,
    /** ISO date `YYYY-MM-DD` (optional; server stores YOB only). */
    val dateOfBirth: String? = null,
    val yob: Int? = null,
    val geoCountry: String? = null,
    val geoRegion: String? = null,
    val interests: List<String> = emptyList(),
    val keywords: List<String> = emptyList(),
    val segments: List<String> = emptyList(),
    val connectionType: String? = null,
    val contentCategory: String? = null,
    val pageType: String? = null,
) {
    fun toUserDataMap(): Map<String, Any> {
        val out = mutableMapOf<String, Any>()
        userPid?.let { out["user_pid"] = it }
        devicePid?.let { out["device_pid"] = it }
        gender?.let { out["gender"] = it }
        age?.let { out["age"] = it }
        val resolvedYob = DemographicsYob.resolveYob(yob, dateOfBirth)
        if (resolvedYob == null && !dateOfBirth.isNullOrBlank()) {
            out["date_of_birth"] = dateOfBirth.trim()
        } else if (resolvedYob != null) {
            out["yob"] = resolvedYob
        }
        geoCountry?.let { out["geo_country"] = it }
        geoRegion?.let { out["geo_region"] = it }
        connectionType?.let { out["connection_type"] = it }
        if (segments.isNotEmpty()) out["segments"] = segments
        return out
    }

    fun toSignalsMap(): Map<String, Any> {
        val base = toUserDataMap().toMutableMap()
        if (interests.isNotEmpty() || keywords.isNotEmpty()) {
            val interestObj = mutableMapOf<String, Any>()
            if (interests.isNotEmpty()) interestObj["tags"] = interests
            if (keywords.isNotEmpty()) interestObj["keywords"] = keywords
            base["interests"] = interestObj
        }
        if (keywords.isNotEmpty()) base["keywords"] = keywords
        contentCategory?.let { base["content_category"] = it }
        pageType?.let { base["page_type"] = it }
        return base
    }

    fun toFirstPartyPayload(os: String, appBundle: String? = null): Map<String, Any> {
        val payload = mutableMapOf<String, Any>(
            "device_pid" to (devicePid ?: ""),
            "os" to os,
        )
        userPid?.let { payload["user_pid"] = it }
        appBundle?.let { payload["app_bundle"] = it }
        if (interests.isNotEmpty() || keywords.isNotEmpty()) {
            payload["interests"] = mapOf(
                "tags" to interests,
                "keywords" to keywords,
            ).filterValues { (it as? List<*>)?.isNotEmpty() == true }
        }
        val meta = mutableMapOf<String, Any>()
        geoCountry?.let { meta["geo_country"] = it }
        val demo = mutableMapOf<String, Any>()
        DemographicsYob.resolveYob(yob, dateOfBirth)?.let { demo["yob"] = it }
        gender?.let { demo["gender"] = it }
        if (demo.isNotEmpty()) meta["demographics"] = demo
        if (meta.isNotEmpty()) payload["metadata"] = meta
        return payload.filterValues { it != "" }
    }
}

File: ios/Sources/DKMadsSSPSDK/Models/TargetingSignals.swift
import Foundation

/// Structured publisher targeting for bid `signals` and optional FPD sync.
public struct TargetingSignals {
    public var userPid: String?
    public var devicePid: String?
    public var gender: String?
    public var age: Int?
    /// ISO `YYYY-MM-DD` (optional; server stores YOB only).
    public var dateOfBirth: String?
    public var yob: Int?
    public var geoCountry: String?
    public var geoRegion: String?
    public var interests: [String]
    public var keywords: [String]
    public var segments: [String]
    public var connectionType: String?
    public var contentCategory: String?
    public var pageType: String?

    public init(
        userPid: String? = nil,
        devicePid: String? = nil,
        gender: String? = nil,
        age: Int? = nil,
        dateOfBirth: String? = nil,
        yob: Int? = nil,
        geoCountry: String? = nil,
        geoRegion: String? = nil,
        interests: [String] = [],
        keywords: [String] = [],
        segments: [String] = [],
        connectionType: String? = nil,
        contentCategory: String? = nil,
        pageType: String? = nil
    ) {
        self.userPid = userPid
        self.devicePid = devicePid
        self.gender = gender
        self.age = age
        self.dateOfBirth = dateOfBirth
        self.yob = yob
        self.geoCountry = geoCountry
        self.geoRegion = geoRegion
        self.interests = interests
        self.keywords = keywords
        self.segments = segments
        self.connectionType = connectionType
        self.contentCategory = contentCategory
        self.pageType = pageType
    }

    public func toUserData() -> UserData {
        var data: UserData = [:]
        if let userPid, !userPid.isEmpty { data["user_pid"] = userPid }
        if let devicePid, !devicePid.isEmpty { data["device_pid"] = devicePid }
        if let gender, !gender.isEmpty { data["gender"] = gender }
        if let age { data["age"] = age }
        if let resolved = DemographicsYob.resolveYob(yob: yob, dateOfBirth: dateOfBirth) {
            data["yob"] = resolved
        } else if let dateOfBirth, !dateOfBirth.isEmpty {
            data["date_of_birth"] = dateOfBirth
        }
        if let geoCountry, !geoCountry.isEmpty { data["geo_country"] = geoCountry }
        if let geoRegion, !geoRegion.isEmpty { data["geo_region"] = geoRegion }
        if let connectionType, !connectionType.isEmpty { data["connection_type"] = connectionType }
        if !segments.isEmpty { data["segments"] = segments }
        return data
    }

    public func toSignalsDictionary() -> [String: Any] {
        var signals = toUserData()
        if !interests.isEmpty || !keywords.isEmpty {
            var interestObj: [String: Any] = [:]
            if !interests.isEmpty { interestObj["tags"] = interests }
            if !keywords.isEmpty { interestObj["keywords"] = keywords }
            signals["interests"] = interestObj
        }
        if !keywords.isEmpty { signals["keywords"] = keywords }
        if let contentCategory, !contentCategory.isEmpty { signals["content_category"] = contentCategory }
        if let pageType, !pageType.isEmpty { signals["page_type"] = pageType }
        return signals
    }

    public func toFirstPartyPayload(os: String, appBundle: String? = nil) -> [String: Any] {
        var payload: [String: Any] = ["os": os]
        if let devicePid, !devicePid.isEmpty { payload["device_pid"] = devicePid }
        if let userPid, !userPid.isEmpty { payload["user_pid"] = userPid }
        if let appBundle, !appBundle.isEmpty { payload["app_bundle"] = appBundle }
        if !interests.isEmpty || !keywords.isEmpty {
            payload["interests"] = [
                "tags": interests,
                "keywords": keywords,
            ]
        }
        var meta: [String: Any] = [:]
        if let geoCountry, !geoCountry.isEmpty { meta["geo_country"] = geoCountry }
        var demo: [String: Any] = [:]
        if let resolved = DemographicsYob.resolveYob(yob: yob, dateOfBirth: dateOfBirth) {
            demo["yob"] = resolved
        }
        if let gender, !gender.isEmpty { demo["gender"] = gender }
        if !demo.isEmpty { meta["demographics"] = demo }
        if !meta.isEmpty { payload["metadata"] = meta }
        return payload
    }
}
